digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/CommonHistoryFiller.scala

```diff
@@ -46,6 +42,34 @@ object CommonHistoryFiller {
     }.toSeq ++ newRoadAddresses.filterNot(ra => unchangedLinks.map(_.roadAddressId).contains(ra.id))
   }
 
+  private def applyTerminated(currentRoadAddresses: Seq[RoadAddress])(projectLinks: Seq[ProjectLink], newRoadAddresses: Seq[RoadAddress]): Seq[RoadAddress] = {
+    val terminatedLinks = projectLinks.filter(_.status == LinkStatus.Terminated)
+
+    terminatedLinks.groupBy(pl => (pl.roadNumber, pl.roadPartNumber, pl.track, pl.roadType)).flatMap {
+      case ((roadNumber, roadPartNumber, trackCode, roadType), groupedLinks) =>
+        val roadAddressesToReturn = newRoadAddresses.filter(ra => groupedLinks.sortBy(_.startAddrMValue).map(_.roadAddressId).contains(ra.id))

```

Comment:
Why groupedLinks needs to be sorted here? Also .map(_.roadAddressId).contains(ra.id) loops through the array twice. It can be refactored like this: .exists(_.roadAddressId == ra.id). This will loop through the array only once and thus the time used for this operation is reduced to half.

Reply:
We need to sort the links because we validate the length of the segment on this method, and it is easier if they are sorted. But you have a point on the loops, I will change those

Reply:
groupedLinks array doesn't keep the changes what sortBy does here. It creates kind of new array when sorted and then calls map to that array. groupedLinks is still unsorted. Might be that it is sorted already in the beginning and that's why this code works, but that sortBy doesn't affect to the rest of the method at all.